### PR TITLE
Tell the user (and report to Sentry) when things break due to WebSockets

### LINKED ISSUE #1644: Set user expectations around WebSockets
### Motivation
When a user's system doesn't support WebSockets, or they've got a firewall, we want them to know that the app won't work for them, rather than throw them into the app and wish 'em the best.

### Parameters
- Easiest path to delivery is just popping a toast.

### Acceptance Criteria
- The user may not enter the application if their system doesn't support WSS
- Send an event to segment so we can track how many of our users this impacts over time

- **Estimated effort:** 12 points ([see CONTRIBUTING.md](https://github.com/ParabolInc/action/blob/master/CONTRIBUTING.md#points-and-sizes))

### PR DESCRIPTION
**Pull request by dan-f**
Fixes #1644 

This PR aims to properly set user expectations when `WebSockets` don't work.

I've included the initial work for review, and am seeking additional feedback in some particular UX areas:
- The language/messaging
- Whether we can come up with a call to action so the user doesn't feel stranded

I'll add some tests as I go along. If you wish to locally test this, the easiest way I found is to stop the server and see how the client responds.

### CODE REVIEW

File: src/client/containers/SocketHealthMonitor/SocketHealthMonitor.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      'WebSocket authentication error',
+      authError
+    );
+  }
+  return <div />;
+};
+
+const mapStateToProps = (state) => {
+  const { socket } = state;
+  if (socket) {
```
[the remark this answers is not in the document]
Comment (dan-f) [reply src/client/containers/SocketHealthMonitor/SocketHealthMonitor.js]: Good to know!

**Comment by dan-f** [comment]
Thanks @mattkrick!

Even if that socketcluster redux library is going away, I do like the idea of keeping track of the websocket connection state in redux; if we rely on the `try/catch` approach, we'd either have to:

- Assume that if a WebSocket connection were to fail, the first connection attempt would be the one to fail
- Poll the server with new WebSocket connections periodically to test health

I personally like the redux approach! All we have to do is add our own actions/reducer once we chuck the socketcluster client lib, right?

**Comment by mattkrick** [comment]
ohhh the redux approach is clunky as heck. i'd equate it to storing the url in redux (which turned out to be a famously bad idea, all of us used to do that, too). in principal, it seems great to keep all the state in 1 place. in practice, the state still lives in the url bar/socket and you're just trying to catch every event that could mutate your state to update redux. if you miss an event, then your redux state isn't accurate. so let's say an error event comes in. are you in an error state? what event clears that error state? more relevant is the connection state. in order to capture the connecting -> connected state, you need the reducer to be the first to instantiate the connection (which is a side effect & a big no no in redux land). While it looks clean & elegant, our current solution is held together by duct tape & pixie dust to ensure that the reducer is the first (but that's all hidden away in the redux-socket-cluster package, so at least our repo looks clean :wink:).

tl;dr let's keep the initial socket connection in our app & shrink the reducer to just be the connection status. subs, errors, etc. get buggy

**Comment by dan-f** [comment]
Hey @mattkrick , I reworked this to account for the fickle state changes occurring in `redux-socket-cluster`, and I believe that duplicate toast issue should be resolved.

File: src/universal/modules/toast/containers/Toast/Toast.js [hunk 2]
```diff
@@ -18,42 +19,47 @@ export default class Toast extends React.Component {
...
   constructor(props) {
     super(props);
-    this.state = {
-      maxNid: 0
-    };
     this.el = null;
   }
 
+  state = {
+    toastToNotification: Immutable.Map() // toast IDs (number) -> 'react-notification-system' IDs (number)
```
[the remark this answers is not in the document]
Comment (dan-f) [reply src/universal/modules/toast/containers/Toast/Toast.js]: I'm well-acquainted with immutable data structures, including those implemented in `immutable-js`, and I disagree with several of your points 😛 

Can I first ask if this is a blocker for you? Keep in mind we're talking about toasts, so like at most ~7 objects in memory at a time.

[the remark this answers is not in the document]
Comment (jordanh) [reply src/universal/modules/toast/containers/Toast/Toast.js]: FWIW @dan-f, I've seen the dragons @mattkrick is talking about. Suggest paths forward (@dan-f I'll leave it to you to pick one):

1) Refactor this now to POJOs or and merge or,
2) Merge this now but create an issue to refactor this in the future

Either is ok by me

[the remark this answers is not in the document]
Comment (jordanh) [reply src/universal/modules/toast/containers/Toast/Toast.js]: Rationale: done is better than perfect for now

[the remark this answers is not in the document]
Comment (dan-f) [reply src/universal/modules/toast/containers/Toast/Toast.js]: I can make the change now, cause it's super small. I'm just puzzled by the rationale (I've used immutable-js at the view layer too!), as this is never passed down to other component instances.

[the remark this answers is not in the document]
Comment (dan-f) [reply src/universal/modules/toast/containers/Toast/Toast.js]: Looking at this again and remembering a few pertinent details:

- This piece of state is used for joining notification data with the `react-notification-system` notification objects, a sadly necessary strategy for dealing with their imperative API
- It should not have anything to do with triggering a render, so I'm taking it out of `this.state` and putting it on `this`.
- It maps number -> number, so POJOs won't do (no support for string keys, and I don't want to coerce strings to numbers)
- So we're now taking about ES6 `Map` vs Immutable-js `Map`. I've replaced `Immutable.Map` with plain `Map` 🤷‍♂️ 

All good?
